File: 3_longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
         if (s == null || s.length() == 0){
            return 0;
        } 
        int i = 0, j = 0, len = s.length();
        Set<Character> set = new HashSet<>();
        int max = 0;
        while (i < len){
            if (!set.contains(s.charAt(i))){
                set.add(s.charAt(i));
                i++;
                max = Math.max(max, set.size());
            }else{
                set.remove(s.charAt(j));
                j++;
            }
        }
        return max;
    }
}

File: 3_longest-substring-without-repeating-characters.py
class Solution:
#     Time: O(n), Space: O(min(n,m)): n(size of string), m(size of set)

    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s or len(s) == 0:
            return 0
        
        seen: Dict[char, int] = {}
        start: int = 0
        max_len: int = 0
        
        for i, char in enumerate(s):
            if char in seen and start <= seen[char]:
                start = seen[char] + 1
            else:
                max_len = max(max_len, i - start + 1)
            
            seen[char] = i
        
        return max_len